documentation/projects/proposals/search_relevancy_sandbox/20230518-implementation_plan_staging_index_rapid_iteration.md

```diff
@@ -0,0 +1,380 @@
+   currently underway. This can be done by checking whether the
+   `{media_type}_data_refresh` and `create_filtered_{media_type}_index` DAGs are
+   currently running. If either of them are, this DAG should fail immediately
+   with an appropriate error message.
+2. Create the index configuration. If `override_config` is supplied, the
+   `index_config` parameter will be used as the entire configuration. If not,
+   the `index_config` parameter will be merged with the existing index settings.
+   See [Merging policy](#merging-policy) for how this merge will be performed.
+   Below are the steps for gathering the current index settings:
+   1. Get the current index information. This will be done using the
+      [`ElasticsearchPythonHook`](https://airflow.apache.org/docs/apache-airflow-providers-elasticsearch/stable/_api/airflow/providers/elasticsearch/hooks/elasticsearch/index.html#airflow.providers.elasticsearch.hooks.elasticsearch.ElasticsearchPythonHook),
+      specifically the
+      [`index.get` function](https://elasticsearch-py.readthedocs.io/en/v8.8.0/api.html#elasticsearch.client.IndicesClient.get).
+   2. Extract the relevant settings from the response. They come back in a form
+      that differs from what is rendered by the
+      [`es_mapping.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/es_mapping.py)
+      file which generates the index settings on the ingestion server.
+      `es_mapping::index_settings` returns data of the form:
+      ```json
+      {
+        "settings": {
+          "index": {
+            "number_of_shards": 18,
+            "number_of_replicas": 0,
+            "refresh_interval": "-1"
+          },
+          "analysis": {...},
+        },
+        "mapping": {...}
+      }
+      ```
+      However, the `index.get` function returns data of the form:
+      ```json
+      {
+        "<index-name>": {
+          "settings": {
+            "index": {
+              "number_of_shards": "18",
+              "number_of_replicas": 0,
+              "refresh_interval": "-1",
+              ...
+              "analysis": {
+                ...
+              }
+            }
+          },
+          "mappings": {...}
+        }
+      }
+      ```
+      The returned information will need to be extracted and converted into the
+      form returned by `es_mapping::index_settings` before it is merged with the
+      new configuration.
+3. Create the new index with the configuration generated in step 1. The index
+   name will either be a combination of the `media_type` and `index_suffix`
+   parameters (`{media_type}-{index_suffix}`, e.g. `image-my-special-suffix`),
+   or the `media_type` and a timestamp (`{media_type}-{timestamp}`, e.g.
+   `image-20200102030405`).
+   ([example from the existing ingestion server code](https://github.com/WordPress/openverse/blob/3fcce5ade2165955db5bbcb4f679257b3260547b/ingestion_server/ingestion_server/indexer.py#L518-L521)).
```

Comment:
Just to confirm (maybe this should be explicit in this IP), this will be done by making the create index request directly to ES, right? In other words, we're not creating the index by making a request to the ingestion server. The links to the ingestion server make this slightly confusing, even though I think I understand they're just meant to be there as examples of how to make the requests using the Python ES client :confused:.

Does the Airflow provider give the same ES client that the ingestion server uses? Otherwise the code samples will be fully irrelevant. Maybe somewhere should specify that we're going to use the `ElasticsearchPythonHook` specifically rather than the `ElasticsearchSQLHook` and then draw the specific connection to the fact that the Python hook returns the same ES client that is used in the ingestion server. That would help explain the use of the ingestion server as code examples.

Reply:
You're correct that we **will not** be using the ingestion server for this operation. The code example that is shared here is merely an example of how the ingestion server does this using the Elasticsearch Python API - I find examples within our own codebase, which have context I'm personally familiar with, more helpful than exclusively linking the Elasticsearch API documentation. 

I also do mention that we will be using the `ElasticsearchPythonHook` for these operations above this: https://github.com/WordPress/openverse/pull/2133/files/65d4e84f8b6570856d8976bee1a364d99c07b82a#diff-344cb6ad9993d7e365933d71faa76a18fbbb062a333a5ae4f73b9dee861b46d8R148 (Line 148 in the current iteration, or step 2.1). I will also add a link to the accompanying Elasticsearch documentation for the `index.create` call here, like I had done in the `index.get` call above.